locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/AbstractComputeServiceRegistry.java

```diff
@@ -0,0 +1,312 @@
+
+        public ComputeServiceSupplierImpl(ConfigBag conf, Iterable<? extends Module> modules, Properties properties) {
+            super(conf, modules, properties);
+        }
+    }
+
+    public class ReusableComputeServiceSupplier extends ComputeServiceSupplier {
+
+        private Map<?, ?> cacheKey;
+
+        public ReusableComputeServiceSupplier(ConfigBag conf, Iterable<? extends Module> modules, Properties properties) {
+            super(conf, modules, properties);
+            this.cacheKey = makeCacheKey();
+        }
+
+        @Override
+        public ComputeService get() {
+            ComputeService result = cachedComputeServices.get(cacheKey);
+            if (result != null) {
+                LOG.trace("jclouds ComputeService cache hit for compute service, for " + Sanitizer.sanitize(getProperties()));
+                return result;
+            }
+            LOG.debug("jclouds ComputeService cache miss for compute service, creating, for " + Sanitizer.sanitize(getProperties()));
+            final ComputeService computeService = super.get();
+            synchronized (cachedComputeServices) {
+                result = cachedComputeServices.get(cacheKey);
+                if (result != null) {
+                    LOG.debug("jclouds ComputeService cache recovery for compute service, for " + Sanitizer.sanitize(cacheKey));
+                    //keep the old one, discard the new one
+                    computeService.getContext().close();
+                    return result;
+                }
+                LOG.debug("jclouds ComputeService created " + computeService + ", adding to cache, for " + Sanitizer.sanitize(getProperties()));
+                cachedComputeServices.put(cacheKey, computeService);
+            }
+            return result;
+        }
+
+        private Map<?, ?> makeCacheKey() {
+            String provider = getProviderFromConfig(getConf());
+            String identity = checkNotNull(getConf().get(CloudLocationConfig.ACCESS_IDENTITY), "identity must not be null");
+            String credential = checkNotNull(getConf().get(CloudLocationConfig.ACCESS_CREDENTIAL), "credential must not be null");
+            String endpoint = getProperties().getProperty(Constants.PROPERTY_ENDPOINT);
+            return MutableMap.builder()
+                    .putAll(getProperties())
+                    .put("provider", provider)
+                    .put("identity", identity)
+                    .put("credential", credential)
+                    .putIfNotNull("endpoint", endpoint)
+                    .build()
+                    .asUnmodifiable();
+        }
+    }
+
+    protected String getProviderFromConfig(ConfigBag conf) {
+        String rawProvider = checkNotNull(conf.get(CLOUD_PROVIDER), "provider must not be null");
+        return DeserializingJcloudsRenamesProvider.INSTANCE.applyJcloudsRenames(rawProvider);
+    }
+
+    private String addEndpointProperty(ConfigBag conf, Properties properties) {
```

Comment:
[minor] maybe this method is can be splitted into:
- getEndpointFromConfig(conf)
- setEndpoint(properties)
?